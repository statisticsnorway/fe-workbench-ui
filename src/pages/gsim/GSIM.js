import React, { Component } from 'react'
import { Link, Route, Switch } from 'react-router-dom'
import { Dropdown, Grid, Icon, Label, Menu, Message } from 'semantic-ui-react'
import { SchemaHandler, UIFormBuilder, UITableBuilder } from 'react-components-library'

import { UI } from '../../utilities/Enum'
import { extractName, handleRoute, splitOnUppercase } from '../../utilities/Common'

const HEADER = {
  en: 'GSIM domains',
  nb: 'GSIM domener'
}

class GSIM extends Component {
  constructor (props) {
    super(props)

    this.state = {
      ready: false,
      schemas: [],
      message: '',
      search: ''
    }
  }

  componentDidMount () {
    const {producer, endpoint, namespace, route} = this.props
    const updatedUrl = endpoint + handleRoute(namespace) + '?schema=embed'

    SchemaHandler(updatedUrl, producer, endpoint, namespace, false, route).then(schemas => {
      this.setState({
        schemas: schemas,
        ready: true
      })
    }).catch(error => {
      this.setState({
        schemas: [],
        ready: true,
        message: error.toString()
      })
    })
  }

  handleSearch = (event) => {
    this.setState({search: event.target.value})
  }

  render () {
    const {ready, schemas, message, search} = this.state
    const {producer, route, endpoint, user, languageCode, namespace} = this.props

    return (
      <Grid>
        <Grid.Column width={2}>
          <Menu vertical fluid>
            <Menu.Item header>
              {HEADER[languageCode]}
              <Label color='teal' size='large'>{ready ? schemas.length : <Icon fitted loading name='spinner' />}</Label>
            </Menu.Item>
            <Dropdown item scrolling search icon='search' placeholder={UI.SEARCH[languageCode]} value={search}
                      onSearchChange={this.handleSearch}>
              <Dropdown.Menu>
                {ready && search === '' && schemas.map((schema, index) => {
                  const domain = extractName(schema.$ref)
                  const link = route + domain

                  return <Dropdown.Item key={index} as={Link} to={link} content={splitOnUppercase(domain)} />
                })}

                {ready && search !== '' && schemas.map((schema, index) => {
                  const domain = extractName(schema.$ref)
                  const link = route + domain

                  if (domain.toUpperCase().includes(search.toUpperCase())) {
                    return <Dropdown.Item key={index} as={Link} to={link} content={splitOnUppercase(domain)} />
                  }

                  return null
                })}
              </Dropdown.Menu>
            </Dropdown>
          </Menu>
        </Grid.Column>
        <Grid.Column width={14}>
          <Switch>
            {ready && message !== '' && <Message error content={message} />}
            {ready && schemas.map((schema, index) => {
              const domain = extractName(schema.$ref)
              const path = route + domain + '/:id'

              return <Route key={index} path={path} exact
                            render={({match}) => <UIFormBuilder params={match.params} producer={producer}
                                                                schema={JSON.parse(JSON.stringify(schema))}
                                                                languageCode={languageCode} namespace={namespace}
                                                                endpoint={endpoint} user={user} />} />
            })}
            {ready && schemas.map((schema, index) => {
              const domain = extractName(schema.$ref)
              const path = route + domain

              return <Route key={index} path={path} exact
                            render={({match}) => <UITableBuilder params={match.params} producer={producer}
                                                                 schema={JSON.parse(JSON.stringify(schema))}
                                                                 languageCode={languageCode} namespace={namespace}
                                                                 endpoint={endpoint} routing={path} />} />
            })}
          </Switch>
        </Grid.Column>
      </Grid>
    )
  }
}

export default GSIM
