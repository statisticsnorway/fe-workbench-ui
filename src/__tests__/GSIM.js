import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'

import GSIM from '../pages/gsim/GSIM'

describe('GSIM', () => {
  it('Search input updates state and value', async () => {
    const component = mount(
      <MemoryRouter>
        <GSIM />
      </MemoryRouter>
    )

    component.find('input').simulate('change', {target: {value: 'Search'}})

    const innerComponent = component.find(GSIM)

    expect(component.find('input').prop('value')).toEqual('Search')
    expect(innerComponent.state('search')).toEqual('Search')
  })
})
