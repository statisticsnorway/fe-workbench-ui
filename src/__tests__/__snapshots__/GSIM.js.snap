// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GSIM Renders itself without error and only once 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GSIM
    endpoint="http://localhost:9090/"
    languageCode="en"
    producer="GSIM"
    route="/home/gsim/"
    user="Test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <GridColumn
          width={2}
        >
          <Menu
            fluid={true}
            vertical={true}
          >
            <MenuItem
              header={true}
            >
              GSIM domains
              <Label
                color="teal"
                size="large"
              >
                <Icon
                  as="i"
                  fitted={true}
                  loading={true}
                  name="spinner"
                />
              </Label>
            </MenuItem>
            <Dropdown
              additionLabel="Add "
              additionPosition="top"
              closeOnBlur={true}
              deburr={false}
              icon="search"
              item={true}
              minCharacters={1}
              noResultsMessage="No results found."
              onSearchChange={[Function]}
              openOnFocus={true}
              placeholder="Search..."
              renderLabel={[Function]}
              scrolling={true}
              search={true}
              searchInput="text"
              selectOnBlur={true}
              selectOnNavigation={true}
              value=""
              wrapSelection={true}
            >
              <DropdownMenu />
            </Dropdown>
          </Menu>
        </GridColumn>,
        <GridColumn
          width={14}
        >
          <Switch />
        </GridColumn>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Menu
            fluid={true}
            vertical={true}
          >
            <MenuItem
              header={true}
            >
              GSIM domains
              <Label
                color="teal"
                size="large"
              >
                <Icon
                  as="i"
                  fitted={true}
                  loading={true}
                  name="spinner"
                />
              </Label>
            </MenuItem>
            <Dropdown
              additionLabel="Add "
              additionPosition="top"
              closeOnBlur={true}
              deburr={false}
              icon="search"
              item={true}
              minCharacters={1}
              noResultsMessage="No results found."
              onSearchChange={[Function]}
              openOnFocus={true}
              placeholder="Search..."
              renderLabel={[Function]}
              scrolling={true}
              search={true}
              searchInput="text"
              selectOnBlur={true}
              selectOnNavigation={true}
              value=""
              wrapSelection={true}
            >
              <DropdownMenu />
            </Dropdown>
          </Menu>,
          "width": 2,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <MenuItem
                header={true}
              >
                GSIM domains
                <Label
                  color="teal"
                  size="large"
                >
                  <Icon
                    as="i"
                    fitted={true}
                    loading={true}
                    name="spinner"
                  />
                </Label>
              </MenuItem>,
              <Dropdown
                additionLabel="Add "
                additionPosition="top"
                closeOnBlur={true}
                deburr={false}
                icon="search"
                item={true}
                minCharacters={1}
                noResultsMessage="No results found."
                onSearchChange={[Function]}
                openOnFocus={true}
                placeholder="Search..."
                renderLabel={[Function]}
                scrolling={true}
                search={true}
                searchInput="text"
                selectOnBlur={true}
                selectOnNavigation={true}
                value=""
                wrapSelection={true}
              >
                <DropdownMenu />
              </Dropdown>,
            ],
            "fluid": true,
            "vertical": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "GSIM domains",
                  <Label
                    color="teal"
                    size="large"
                  >
                    <Icon
                      as="i"
                      fitted={true}
                      loading={true}
                      name="spinner"
                    />
                  </Label>,
                ],
                "header": true,
              },
              "ref": null,
              "rendered": Array [
                "GSIM domains",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": <Icon
                      as="i"
                      fitted={true}
                      loading={true}
                      name="spinner"
                    />,
                    "color": "teal",
                    "size": "large",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "as": "i",
                      "fitted": true,
                      "loading": true,
                      "name": "spinner",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "additionLabel": "Add ",
                "additionPosition": "top",
                "children": <DropdownMenu />,
                "closeOnBlur": true,
                "deburr": false,
                "icon": "search",
                "item": true,
                "minCharacters": 1,
                "noResultsMessage": "No results found.",
                "onSearchChange": [Function],
                "openOnFocus": true,
                "placeholder": "Search...",
                "renderLabel": [Function],
                "scrolling": true,
                "search": true,
                "searchInput": "text",
                "selectOnBlur": true,
                "selectOnNavigation": true,
                "value": "",
                "wrapSelection": true,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    false,
                    false,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  false,
                  false,
                ],
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Switch />,
          "width": 14,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              false,
              false,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            false,
            false,
            false,
          ],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <GridColumn
            width={2}
          >
            <Menu
              fluid={true}
              vertical={true}
            >
              <MenuItem
                header={true}
              >
                GSIM domains
                <Label
                  color="teal"
                  size="large"
                >
                  <Icon
                    as="i"
                    fitted={true}
                    loading={true}
                    name="spinner"
                  />
                </Label>
              </MenuItem>
              <Dropdown
                additionLabel="Add "
                additionPosition="top"
                closeOnBlur={true}
                deburr={false}
                icon="search"
                item={true}
                minCharacters={1}
                noResultsMessage="No results found."
                onSearchChange={[Function]}
                openOnFocus={true}
                placeholder="Search..."
                renderLabel={[Function]}
                scrolling={true}
                search={true}
                searchInput="text"
                selectOnBlur={true}
                selectOnNavigation={true}
                value=""
                wrapSelection={true}
              >
                <DropdownMenu />
              </Dropdown>
            </Menu>
          </GridColumn>,
          <GridColumn
            width={14}
          >
            <Switch />
          </GridColumn>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Menu
              fluid={true}
              vertical={true}
            >
              <MenuItem
                header={true}
              >
                GSIM domains
                <Label
                  color="teal"
                  size="large"
                >
                  <Icon
                    as="i"
                    fitted={true}
                    loading={true}
                    name="spinner"
                  />
                </Label>
              </MenuItem>
              <Dropdown
                additionLabel="Add "
                additionPosition="top"
                closeOnBlur={true}
                deburr={false}
                icon="search"
                item={true}
                minCharacters={1}
                noResultsMessage="No results found."
                onSearchChange={[Function]}
                openOnFocus={true}
                placeholder="Search..."
                renderLabel={[Function]}
                scrolling={true}
                search={true}
                searchInput="text"
                selectOnBlur={true}
                selectOnNavigation={true}
                value=""
                wrapSelection={true}
              >
                <DropdownMenu />
              </Dropdown>
            </Menu>,
            "width": 2,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <MenuItem
                  header={true}
                >
                  GSIM domains
                  <Label
                    color="teal"
                    size="large"
                  >
                    <Icon
                      as="i"
                      fitted={true}
                      loading={true}
                      name="spinner"
                    />
                  </Label>
                </MenuItem>,
                <Dropdown
                  additionLabel="Add "
                  additionPosition="top"
                  closeOnBlur={true}
                  deburr={false}
                  icon="search"
                  item={true}
                  minCharacters={1}
                  noResultsMessage="No results found."
                  onSearchChange={[Function]}
                  openOnFocus={true}
                  placeholder="Search..."
                  renderLabel={[Function]}
                  scrolling={true}
                  search={true}
                  searchInput="text"
                  selectOnBlur={true}
                  selectOnNavigation={true}
                  value=""
                  wrapSelection={true}
                >
                  <DropdownMenu />
                </Dropdown>,
              ],
              "fluid": true,
              "vertical": true,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    "GSIM domains",
                    <Label
                      color="teal"
                      size="large"
                    >
                      <Icon
                        as="i"
                        fitted={true}
                        loading={true}
                        name="spinner"
                      />
                    </Label>,
                  ],
                  "header": true,
                },
                "ref": null,
                "rendered": Array [
                  "GSIM domains",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": <Icon
                        as="i"
                        fitted={true}
                        loading={true}
                        name="spinner"
                      />,
                      "color": "teal",
                      "size": "large",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "class",
                      "props": Object {
                        "as": "i",
                        "fitted": true,
                        "loading": true,
                        "name": "spinner",
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "additionLabel": "Add ",
                  "additionPosition": "top",
                  "children": <DropdownMenu />,
                  "closeOnBlur": true,
                  "deburr": false,
                  "icon": "search",
                  "item": true,
                  "minCharacters": 1,
                  "noResultsMessage": "No results found.",
                  "onSearchChange": [Function],
                  "openOnFocus": true,
                  "placeholder": "Search...",
                  "renderLabel": [Function],
                  "scrolling": true,
                  "search": true,
                  "searchInput": "text",
                  "selectOnBlur": true,
                  "selectOnNavigation": true,
                  "value": "",
                  "wrapSelection": true,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      false,
                      false,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    false,
                    false,
                  ],
                  "type": [Function],
                },
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Switch />,
            "width": 14,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                false,
                false,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              false,
              false,
              false,
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
